daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala

```diff
@@ -65,123 +79,70 @@ case object VariableInProcess extends VariableState
-   *
-   * This function allows state to be set while also enabling stateArg to be
-   * lazily evaluated. This allows the VariableInstance to support setting the
-   * state later and allows the construction of the VariableMap containing the
-   * VariableInstance before evaluating the default value expression
-   */
-  def state = {
-    if (_state == null) {
-      _state = stateArg
-    }
+  var state: VariableState = VariableUndefined
+  var priorState: VariableState = VariableUndefined
+  var value: DataValuePrimitiveNullable = DataValue.NoValue
+  var priorValue: DataValuePrimitiveNullable = DataValue.NoValue
 
-    _state
-  }
+  // This represents the default value at the start of processing, provided by
+  // either the defaultValue expression or by an external binding
+  var firstInstanceInitialValue: DataValuePrimitiveNullable = DataValue.NoValue
 
   def setState(s: VariableState) = {
     this.priorState = this.state
-    this._state = s
+    this.state = s
   }
 
   def setValue(v: DataValuePrimitiveNullable) = {
     this.priorValue = this.value
-    this._value = Some(v)
+    this.value = v
   }
 
-  def setDefaultValue(v: DataValuePrimitiveNullable) = this._value = Some(v)
-
-  override def preSerialization: Unit = {
-    defaultValueExpr
-    value
-    state
+  /* This is used to set a default value without assigning a priorValue */
+  def setDefaultValue(v: DataValuePrimitiveNullable) = {
+    Assert.invariant((this.state == VariableUndefined || this.state == VariableInProcess) && v.isDefined)
+    this.state = VariableDefined
+    this.value = v
   }
 
-  override def toString: String = "VariableInstance(%s,%s,%s,%s,%s,%s,%s)".format(
+  override def toString: String = "VariableInstance(%s,%s,%s,%s,%s,%s)".format(
                                                     state,
                                                     value,
                                                     rd,
-                                                    defaultValueExpr,
+                                                    rd.maybeDefaultValueExpr,
                                                     priorState,
-                                                    priorValue,
-                                                    this.hashCode)
+                                                    priorValue)
 
   def copy(
     state: VariableState = state,
     value: DataValuePrimitiveNullable = value,
     rd: VariableRuntimeData = rd,
-    defaultValueExpr: Maybe[CompiledExpression[AnyRef]] = defaultValueExpr,
     priorState: VariableState = priorState,
     priorValue: DataValuePrimitiveNullable = priorValue) = {
-      val inst = new VariableInstance(state, value, rd, defaultValueExpr, priorState, priorValue)
-      inst.init()
+      val inst = new VariableInstance(rd)
+      inst.state = state
+      inst.priorState = priorState
+      inst.value = value
+      inst.priorValue = priorValue
       inst
```

Comment:
Now that we no onger need init(), I wonder if we should just make ``VariableInstance`` a case class now? Then we get this copy function for free and Scala ensures all state is copied (in fact, right now, we aren't copying firstInstanceInitialValue so there's potentially a bug). That also means we can get rid of the ``VariableInstance`` companion object since case class will also give us that for free.

Reply:
Done

Reply:
Actually, I dug into this a bit more and it seems that when using a case class the copy method is only a shallow copy.  This ends up causing several tests to break (weirdly none in the TestVariables TDML tests, which is why I thought this was an easy change).  I've reverted back to using a normal class again.

Reply:
Just curious, what in VariableInstance needs a deep copy? Isn't all the state just enums and things that don't have a differencne between deep/shallow copy? Wonder if something else is going on?

Reply:
Ah, nevermind, the default case class copy() doesn't copy member var's, only copies the constructor parameters. So all our var state isn't copied with a case class. You're right that we still need this copy method.

Reply:
Might be worth copying the firstInstanceInitialValue just for correctness though.